webhooks.md

```diff
@@ -43,6 +43,199 @@ There are also additions/changes made to the `config` directory and to `main.go`
+		return apierrors.NewInvalid(
+			schema.GroupKind{Group: "foo.openstack.org", Kind: "Foo"},
+			r.Name, errors)
+	}
+	return nil
+}
+
+func (r *NovaSpec) ValidateUpdate(old runtime.Object, basePath *field.Path) field.ErrorList {
+    // Implement the create validation here and return a list of errors.
+    var errors field.ErrorList
+    // You can also reuse specific validation from create validation here if
+    // needed
+    errors = append(errors,  r.ValidateMyFieldLength(basePath)...)
+
+	return errors
+}
+
+```
+
+In the openstack-operator
+```golang
+func (r *OpenStackControlPlane) ValidateCreate() error {
+	var allErrs field.ErrorList
+	basePath := field.NewPath("spec")
+	if err := r.ValidateServices(basePath); err != nil {
+		allErrs = append(allErrs, err...)
+	}
+
+	if len(allErrs) != 0 {
+		return apierrors.NewInvalid(
+			schema.GroupKind{Group: "core.openstack.org", Kind: "OpenStackControlPlane"},
+			r.Name, allErrs)
+	}
+
+	return nil
+}
+
+func (r *OpenStackControlPlane) ValidateCreateServices(basePath *field.Path) field.ErrorList {
+    var errors field.ErrorList
+    // We do validation that are not Create and Update specific here
+    errors := append(errors, r.ValidateServices(basePath)...)
+    // But also call the service Spec specific Create validation
+    if r.Spec.Foo.Enabled {
+		errors = append(errors, r.Spec.Foo.Template.ValidateCreate(basePath.Child("foo").Child("template"))...)
+	}
+    return errors
+}
+
+func (r *OpenStackControlPlane) ValidateUpdate(old runtime.Object) error {
+	var allErrs field.ErrorList
+	basePath := field.NewPath("spec")
+	if err := r.ValidateServices(basePath); err != nil {
+		allErrs = append(allErrs, err...)
+	}
+
+	if len(allErrs) != 0 {
+		return apierrors.NewInvalid(
+			schema.GroupKind{Group: "core.openstack.org", Kind: "OpenStackControlPlane"},
+			r.Name, allErrs)
+	}
+
+	return nil
+}
+
+func (r *OpenStackControlPlane) ValidateUpdateServices(old runtime.Object, basePath *field.Path) field.ErrorList {
+    var errors field.ErrorList
+    // We do validation that are not Create and Update specific here
+    errors := append(errors, r.ValidateServices(basePath)...)
+    // But also call the service Spec specific Update validation
+    if r.Spec.Foo.Enabled {
+		errors = append(errors, r.Spec.Foo.Template.ValidateUpdate(old, basePath.Child("foo").Child("template"))...)
+	}
+    return errors
+}
+```
+
+So when the webhook runs in the service operator for a create the call chain
+is:
+1. Foo.ValidateCreate()
+2. FooSpec.ValidateCreate()
+3. FooSpec.ValidateMyFieldLength()
+
+When the webhook runs in the openstack-operator for an update the call chain
+is:
+1. OpenStackControlPlane.ValidateUpdate()
+2. FooSpec.ValidateUpdate()
+3. FooSpec.ValidateMyFieldLength()
+
+This validation chaining patter can be further extended inside the service
+operator if the service operator not just implement a top level service CRD but
+also sub CRDs.
```

Comment:
I agree that we should find a common pattern for the upper layer interfaces, and when we reach the `Validate` function at `FooSpec` level then it's the <service>-operator itself that runs a set of validations based on the managed resources.
For instance I see the `ValidateMyFieldLength` at third level in the call chain above, but for me step 3 represents the Spec layer where `ValidateMyFieldLength` is just one of the set of validations I could run (and it depends on the service CRD and the exposed fields).
@gibizer Is my understanding correct?
Should we clarify that `ValidateMyFieldLength` is an example and from an high level point of view we can have a set of validations at Spec level and the generated resources?

Reply:
Your understanding is correct. ValidateMyFieldLength is an example to show how to reuse a Validation that needs to be done both in case of Create and Update. There could be validations that only happen during Update (i.e. immutability check).

The pattern stops at the FooSpec level. The rest is just an example. I clarified this in the doc now.